Fixes #30444 - Host registration template

stejskalleos opened the pull request:
**Features:**
* Enable build mode for unmanaged hosts.
* Catch errors when rendering Host register template.
* Add register token to host detail in api (so it can be used later in Global registration template)

**Notes:**
* PR is part of **Simple & automatic host registration WF**, [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588) & [Redmine](https://projects.theforeman.org/issues/30440)
* Adding Host registration template to Global registration template will be done in [#30445](https://projects.theforeman.org/issues/30445)

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

stejskalleos committed:
Fixes #30444 - Host registration in Global Registration template

Add Host registration template to the GR template,
enable build mode for unmanaged hosts,
catch errors when rendering HR template and
add register token to the host detail in api.

ares commented on app/controllers/foreman_register/hosts_controller.rb:
```diff
@@ -19,11 +19,20 @@ class HostsController < ::ForemanRegister::ApplicationController
-      template = rendered_registration_template
-      render plain: template if template
+      if @template
+        @host.setBuild(validate_on_save: @host.managed?)
```
why don't we rather disable the validation at https://github.com/theforeman/foreman/blob/2801602393aee2245a2f5066d3dc4e731fea40e2/app/models/host/base.rb#L42

it seems this patch is saying that it's a valid thing to set the build status for unmanaged host. In the RFC thread, I haven't seen any concerns with it, besides suggestion of getting rid of unmanaged host (which is a long term goal). We don't need this argument then at all.

ares commented on test/controllers/foreman_register/hosts_controller_test.rb:
```diff
@@ -75,6 +75,28 @@ class HostsControllerTest < ActionController::TestCase
+        it 'no OS' do
+          get :register, params: { token: FactoryBot.create(:host).registration_token }
+          assert_response :bad_request
+          assert_includes @response.body, "[Foreman::Exception]: Must provide an operating system\"\nexit 1\n"
```
The message is a bit confusing, it suggests user should provide that e.g. as a parameter. It should be saying the host does not have OS associated. I know it's not added in this PR but could we change it in here, before we hardcode it in our test suite?

ares commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -33,6 +33,8 @@ end
 # display the token, if it hasn't expired
 node(:token, :if => ->(h) { h.token && !h.token_expired? }) { |host| host.token.value }
 
+node(:registration_token) { |h| h.registration_token }
```
Should we have a test for this attribute being present? Probably at https://github.com/theforeman/foreman/blob/2801602393aee2245a2f5066d3dc4e731fea40e2/test/controllers/api/v2/hosts_controller_test.rb

ares commented on app/models/host/managed.rb:
```diff
@@ -499,10 +499,10 @@ def populate_fields_from_facts(parser, type, source_proxy)
     self.build = true
     self.initiated_at = Time.now.utc
-    logger.warn("Set build failed: #{errors.inspect}") unless save
+    logger.warn("Set build failed: #{errors.inspect}") unless save(validate: validate_on_save)
```
I think we should add a test that this method also works for unmanaged hosts, meaning it sets the build mode for it

ares reviewed (changes requested):
Good start for sure, left few comments. I'd like to see few more tests and a different approach for builds status for unmanaged hosts.

stejskalleos commented:
Rebased & updated
- added tests & updated old ones
- removed validation for setBuild when host is unmanaged
- Updated error message for host with no OS

lzap commented on app/controllers/foreman_register/hosts_controller.rb:
```diff
@@ -19,11 +19,20 @@ class HostsController < ::ForemanRegister::ApplicationController
-      template = rendered_registration_template
-      render plain: template if template
+      if @template
+        @host.setBuild
```
I am not comfortable with this to be honest. Is this really needed? Leos mentioned that you, @ares, wanted this.

This changes state of the host just by calling register endpoint which is supposed to render a template. This can either cause some orchestration issues (user-based hooks for example) or create an attack vector when an existing (managed) host could be put into build mode (possible data loss after reboot).

If we want the build mode, why don't we call set build explicitly? There is an API endpoint to do that.

Finally, there must was a reason why build validation for unmanaged hosts was added: https://projects.theforeman.org/issues/8587 specifically that the host is in incorrect state (PXE templates were not deployed). When a host is then managed, templates do not get rendered and it's still in incorrect mode.

ares replied:
This PR basically allows unmanaged host to start using build mode signalling. Once the registration template is downloaded, it turns the host to "pending installation". After the registration is complete (can take minutes, as it can install puppet, insights etc), it turns it to "built". I think this is a great benefit comparing not knowing that some host is being registered. This tells you the host is not ready for using yet, even though it exists in the DB (we need to create it before we render the template for it)

> This changes state of the host just by calling register endpoint which is supposed to render a template. This can either cause some orchestration issues (user-based hooks for example)

I'm not concerned with that (in combination with the answer below). Let's assume we can only run the registration for unmanaged hosts. There's no orchestration for unmanaged hosts. I'm not sure about hooks plugin, but I don't think people would define any hooks for unmanaged hosts. They are typically not created from the Foreman, but rather by e.g. puppet.

> or create an attack vector when an existing (managed) host could be put into build mode (possible data loss after reboot).

You're right, we should make sure the `@host` is unmanaged, we shouldn't allow registering managed hosts.

> If we want the build mode, why don't we call set build explicitly? There is an API endpoint to do that.

we can't, the validation prevents unmanaged host to be set to be built, that's the change in this PR.

> Finally, there must was a reason why build validation for unmanaged hosts was added:

We should then perhaps fix this properly and reset the build mode when host is switched between managed - unmanaged. Or rerender the templates. The issue you linked was talking about possibility of setting the build mode for unmanaged host through API/UI. We allow this for unmanaged hosts only via registration process.

stejskalleos replied:
Okay, just to confirm: the decision is to allow build mode **only for unmanaged hosts**

ares replied:
I think the conclusion (I hope others agree with) is, keep removal of the validation so both managed and newly unmanaged host can be turned to build mode. But the `/register` endpoint should verify the specified host is actually unmanaged and fail if it is managed.

ares replied:
As part of testing I figured the behavior described in here was the original intention and we should have this endpoint available for managed hosts too, calling register is similar to switching the host to build, it will clear facts and reports without a warning (through API), but that's by design. Only authorized users can do that.

ekohl commented on app/models/provisioning_template.rb:
```diff
@@ -97,7 +97,7 @@ def clone
   def self.find_template(opts = {})
     raise ::Foreman::Exception.new(N_("Must provide template kind")) unless opts[:kind]
-    raise ::Foreman::Exception.new(N_("Must provide an operating system")) unless opts[:operatingsystem_id]
+    raise ::Foreman::Exception.new(N_("Host is not associated with an operating system")) unless opts[:operatingsystem_id]
```
This method feels like an internal API and it looks like this is returning a user error where it should really return an API error; in other words, solving it at the wrong layer. Here we don't have anything like host, just a set of options. Perhaps the register controller should first ensure an OS is associated before calling this. If not, present the user with the correct error.

The alternative is to do so in the HostExtension. There you can also provide context about which host it is, rather than some unknown host.

stejskalleos commented:
Updated & rebased.

ares commented on app/controllers/foreman_register/hosts_controller.rb:
```diff
@@ -44,31 +54,32 @@ def render_error(message:, status: :bad_request, **kwargs)
+
+    def build_host
+      if @host.managed?
+        render_error(message: N_('Build mode cannot be enabled for managed host'))
```
This is a confusing error message. There's nothing wrong in building the managed host in general, it's only the registration endpoint that can't be used for building managed host. I'd move this logic to the register method, as it's really only specific to it. You'd also got rid of before_action condition that relies on `@template` status. The whole method would be nicer if you use guards. E.g.

```ruby
    def register
      if @template.nil?
        msg = N_("Unable to find registration template for host %{host} running %{os}") % { host: @host.name, os: @host.operatingsystem }
        render_error(
          message: msg,
          status: :not_found,
          host: @host.name,
          os: @host.operatingsystem
        )
        return
      end

      if @host.managed?
          .... handle the error
          return
       end

    @host.setBuild
     safe_render(@template)
  end
```

stejskalleos replied:
And how about not handling the error at all? We can do just:
```
@host.setBuild unless @host.managed?
```

ares reviewed (changes requested):
I think that is my last comment to the code, otherwise looks good. Will proceed to testing right after.

ares reviewed (approved):
Thanks @stejskalleos, all comments resolved. I checked with @lzap offline and he has no more comments. Finally merging! Thanks everyone for comments.

ezr-ondrej commented:
This causes a https://projects.theforeman.org/issues/31518 because it tries to create the token on a new host during the orchestration.
In general I think it is not a very good practice for a getter to be initializing the value.
@stejskalleos would it make sense to return `nil` if the facet doesn't exists here instead of creating it?
https://github.com/theforeman/foreman/blob/bcd650273e744537a40198c937184a04ff1a180b/app/models/concerns/foreman_register/host_extensions.rb#L12

We would probably need to initialize it somewhere else though.
